metric/internal/global/meter_test.go

```diff
@@ -0,0 +1,286 @@
+
+	afcounter, err := m.AsyncFloat64().Counter("test_Async_Counter")
+	require.NoError(t, err)
+	_, err = m.AsyncFloat64().UpDownCounter("test_Async_UpDownCounter")
+	require.NoError(t, err)
+	_, err = m.AsyncFloat64().Gauge("test_Async_Gauge")
+	require.NoError(t, err)
+
+	_, err = m.AsyncInt64().Counter("test_Async_Counter")
+	require.NoError(t, err)
+	_, err = m.AsyncInt64().UpDownCounter("test_Async_UpDownCounter")
+	require.NoError(t, err)
+	_, err = m.AsyncInt64().Gauge("test_Async_Gauge")
+	require.NoError(t, err)
+
+	_ = m.RegisterCallback([]instrument.Asynchronous{afcounter}, func(ctx context.Context) {
+		afcounter.Observe(ctx, 3)
+	})
+
+	sfcounter, err := m.SyncFloat64().Counter("test_Async_Counter")
+	require.NoError(t, err)
+	_, err = m.SyncFloat64().UpDownCounter("test_Async_UpDownCounter")
+	require.NoError(t, err)
+	_, err = m.SyncFloat64().Histogram("test_Async_Histogram")
+	require.NoError(t, err)
+
+	_, err = m.SyncInt64().Counter("test_Async_Counter")
+	require.NoError(t, err)
+	_, err = m.SyncInt64().UpDownCounter("test_Async_UpDownCounter")
+	require.NoError(t, err)
+	_, err = m.SyncInt64().Histogram("test_Async_Histogram")
+	require.NoError(t, err)
+
+	return sfcounter, afcounter
+}
+
+// This is to emulate a read from an exporter
+func testCollect(t *testing.T, m metric.Meter) {
+	if tMeter, ok := m.(*meter); ok {
+		m, ok = tMeter.delegate.Load().(metric.Meter)
+		if !ok {
+			t.Error("meter was not delegated")
+			return
+		}
+	}
+	tMeter, ok := m.(*testMeter)
+	if !ok {
+		t.Error("collect called on non-test Meter")
+		return
+	}
+	tMeter.collect()
+}
+
+func Test_MeterProvider_delegates_calls(t *testing.T) {
+
+	// The global MeterProvider should directly call the underlying MeterProvider
+	// if it is set prior to Meter() being called.
+
+	// globalMeterProvider := otel.GetMeterProvider
+	globalMeterProvider := &meterProvider{}
+
+	mp := &testMeterProvider{}
+
+	// otel.SetMeterProvider(mp)
+	globalMeterProvider.setDelegate(mp)
+
+	require.Equal(t, 0, mp.count)
+
+	meter := globalMeterProvider.Meter("go.opentelemetry.io/otel/metric/internal/global/meter_test")
+
+	ctr, actr := testSetupAllInstrumentTypes(t, meter)
+
+	ctr.Add(context.Background(), 5)
+
+	testCollect(t, meter) // This is a hacky way to emulate a read from an exporter
+
+	// Calls to Meter() after setDelegate() should be executed by the delegate
+	require.IsType(t, &testMeter{}, meter)
+	if tMeter, ok := meter.(*testMeter); ok {
+		require.Equal(t, 3, tMeter.afCount)
+		require.Equal(t, 3, tMeter.aiCount)
+		require.Equal(t, 3, tMeter.sfCount)
+		require.Equal(t, 3, tMeter.siCount)
+		require.Equal(t, 1, len(tMeter.callbacks))
+	}
+
+	// Because the Meter was provided by testmeterProvider it should also return our test instrument
+	require.IsType(t, &testCountingFloatInstrument{}, ctr, "the meter did not delegate calls to the meter")
+	if testCtr, ok := ctr.(*testCountingFloatInstrument); ok {
+		require.Equal(t, 1, testCtr.count)
+	}
```

Comment:
Given the requirement above, having this silently passed over it the type assertion cannot be made will only hide a error. Why not panic if the assertion can not be made to communicate the error to the developer early (they likely made the change to the code that caused the panic).

```suggestion
	assert.Equal(t, 1, ctr.(*testCountingFloatInstrument).count)
```

Reply:
The difference is that the test will fail as written, with the message that it did not delegate the call.  If we were to then cast it to the wrong type after the fail, execution doesn't stop, we would **also** get a panic message, which I don't think is that helpful.

Reply:
The type assertion is required though, the execution would stop. Reaching here will only happen if it is the correct type so it shouldn't panic.

If a future change removed that requirement, as it currently is, a failed assertion would be passed by. The proposed change would only panic in that situation.

Reply:
Agreed we do not want to replace all `required` with `assert`. We should only replace the ones that do not assert things needed for continued execution. That way multiple tested conditions can be reported for a single run (instead of having to fix one and discover the next).